lib/srv/ctx.go

```diff
@@ -584,6 +603,128 @@ func (c *ServerContext) getSession() *session {
 	return c.session
 }
 
+// GetX11Config gets the x11 config for this server session.
+func (c *ServerContext) GetX11Config() *X11Config {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.x11Config
+}
+
+// OpenXServerListener opens a new xserver unix listener.
+func (c *ServerContext) OpenXServerListener(x11Req x11.ForwardRequestPayload, displayOffset int) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.x11Config != nil {
+		return trace.AlreadyExists("x11 forwarding is already set up for this session")
+	}
+
+	l, display, err := x11.OpenNewXServerListener(displayOffset, x11Req.ScreenNumber)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// TODO (Joerger): During a non-graceful shutdown, such as SIGTERM, the Teleport
```

Comment:
This seems worth fixing sooner rather than later.

Reply:
It's worth mentioning that this is also an issue with unix sockets opened for agent forwarding, so I'll work on a fix for non-graceful shutdown's separate from this PR.